source/adapters/cuda/usm.cpp

```diff
@@ -80,54 +91,32 @@ urUSMSharedAlloc(ur_context_handle_t hContext, ur_device_handle_t hDevice,
                 (alignment == 0 || ((alignment & (alignment - 1)) == 0)),
             UR_RESULT_ERROR_INVALID_VALUE);
 
-  if (!hPool) {
-    return USMSharedAllocImpl(ppMem, hContext, hDevice, nullptr, nullptr, size,
-                              alignment);
+  umf_memory_pool_handle_t hPoolInternal = nullptr;
+  usm::pool_descriptor Desc = {hPool, hContext, hDevice, UR_USM_TYPE_SHARED,
+                               false};
+  if (hPool) {
+    auto hPoolInternalOpt = hPool->PoolManager.getPool(Desc);
+    hPoolInternal = hPoolInternalOpt.value();
+  } else {
+    auto hPoolInternalOpt = hContext->ProxyPoolManager.getPool(Desc);
+    hPoolInternal = hPoolInternalOpt.value();
   }
 
-  auto UMFPool = hPool->SharedMemPool.get();
-  *ppMem = umfPoolAlignedMalloc(UMFPool, size, alignment);
+  *ppMem = umfPoolAlignedMalloc(hPoolInternal, size, alignment);
   if (*ppMem == nullptr) {
-    auto umfErr = umfPoolGetLastAllocationError(UMFPool);
+    auto umfErr = umfPoolGetLastAllocationError(hPoolInternal);
     return umf::umf2urResult(umfErr);
   }
   return UR_RESULT_SUCCESS;
 }
 
-ur_result_t USMFreeImpl(ur_context_handle_t Context, void *Pointer) {
-  ur_result_t Result = UR_RESULT_SUCCESS;
-  try {
-    ScopedContext Active(Context);
-    bool IsManaged;
-    unsigned int Type;
-    void *AttributeValues[2] = {&IsManaged, &Type};
-    CUpointer_attribute Attributes[2] = {CU_POINTER_ATTRIBUTE_IS_MANAGED,
```

Comment:
What happened with this code? (CUpointer_attribute, etc.)

Reply:
I divided this into two functions, `USMFreeImpl` and `USMHostFreeImpl`. Since providers allocate specific memory types, I could assign the `cuMemFree` and `cuMemFreeHost` to appropriate providers and avoid calling `cuPointerGetAttributes`.